### ISSUE TITLE
Using Edward in Mixture Density Networks

### ISSUE THREAD

**Issue by cbonnett** [blei-lab/edward#88]
Given that Edward implements many distributions in TF, I thought it would be a natural to use these distributions for Mixture Density Networks (MDN). 
Mixture Density Networks are a type feed forward neural net that make a prediction as mixture of  distributions. See a blog post I wrote about them : http://cbonnett.github.io/MDN.html

The combination of Edward, TF and Keras would make creating these easier. 
There seems to be one roadblock for using Edward for  MDN's and that is the following line in each of the density distributions: 
 
```python
x = tf.cast(tf.squeeze(x), dtype=tf.float32)
```

And in particular the tf.squeeze for x when dealing with a 1-D likelihood (I have yet to try higher dimensions). In Particular, TF complains on the following line : 

```python
z = (x - loc) / scale
``` 
in the ```norm ```function when ```loc``` and ```scale``` have dimension ```[? ,10]``` if we are trying to fit a mixture of 10 normal distrbutions and x has ```[?,1]``` before ```squeeze``` is applied. TF complains about the dimensionality, ```x``` needs that extra dimension for MDN's.  
 
Does Edward break if the squeeze is removed ? (i'm guessing it will)
Any idea for a solution for this ? 

Here is the Full MDN code but without Edward to the above mentioned problem:

```
from keras.layers import Input, Dense, merge
import tensorflow as tf
sess = tf.Session()
from keras import backend as K
K.set_session(sess)

mixture_components = 10
NSAMPLE = 6000

y_data = np.float32(np.random.uniform(-10.5, 10.5, (1, NSAMPLE))).T
r_data = np.float32(np.random.normal(size=(NSAMPLE,1))) # random noise
x_data = np.float32(np.sin(0.75*y_data)*7.0+y_data*0.5+r_data*1.0)

X_train, X_valid, y_train, y_valid = train_test_split(x_data, y_data, random_state=42)

print x_data.shape, X_train.shape
print y_data.shape, y_train.shape

X = tf.placeholder(tf.float32, shape=(None, 1))
y = tf.placeholder(tf.float32, shape=(None, 1))

hidden1 = Dense(25, activation='relu')(X)  # fully-connected layer with 128 units and ReLU activation
hidden2 = Dense(25, activation='relu')(hidden1)
means = Dense(mixture_components)(hidden2)
standard_deviations = Dense(mixture_components, activation=K.exp)(hidden2)
weights = Dense(mixture_components, activation=K.softmax)(hidden2)

def logpdf(x, loc=0, scale=1):
    z = (x - loc) / scale
    return (-0.5*tf.log(2*np.pi) - tf.log(scale) - 0.5*tf.square(z))

def get_lossfunc(weigts, standard_deviations, means, y):
    result = tf.exp(logpdf(y, means, standard_deviations))
    result = tf.mul(result, weights)
    result = tf.reduce_sum(result, 1, keep_dims=True)
    result = -tf.log(result)
    return tf.reduce_mean(result)

print weights.get_shape


lossfunc = get_lossfunc(weights, standard_deviations, means, y)
train_op = tf.train.AdamOptimizer().minimize(lossfunc)

sess = tf.InteractiveSession()
sess.run(tf.initialize_all_variables())

NEPOCH = 20
train_loss = np.zeros(NEPOCH) # store the training progress here.
valid_loss = np.zeros(NEPOCH)
for i in range(NEPOCH):    
    _, train_loss[i] = sess.run([train_op, lossfunc], feed_dict={X: X_train, y: y_train})
    valid_loss[i] = lossfunc.eval(feed_dict={X: X_valid, y: y_valid})
    
    pred_weights, pred_means, pred_std = sess.run([weights, means, standard_deviations], 
                                                  feed_dict={X: X_valid})
````

**Comment by dustinvtran** [comment]
Hi @cbonnett, thanks for the comment. The example looks useful! I changed Edward in the above pull request to let its `norm.logpdf()` work in your script.

I am also re-organizing the example a little so it may be able leverage other stuff Edward also has to offer (e.g., `ed.MAP` for maximum likelihood, although keeping the `for` loop in the script to stay idiomatically TensorFlow). Once I'm done, mind if I showcase this in the [`examples/`](https://github.com/blei-lab/edward/tree/master/examples) directory?

**Comment by cbonnett** [comment]
Hi @dustinvtran ,

That's great news.  Feel free to use it in examples. 
I wanted to write a notebook/blogpost showing how to use Edward, Keras and TF to construct MDN's. 
this would fit with  issue #35  to make some notebooks like the gpy library.  I'd be happy to contribute to that.

**Comment by cbonnett** [comment]
Great. I'll wait untill pull request #90 has been merged.

**Comment by dustinvtran** [comment]
@cbonnett: Done!

**Comment by cbonnett** [comment]
I've got a decent draft of the blogpost which written in a notebook, so can be easily adjusted to serve as a tutorial notebook. I gave you access to the repo. (https://github.com/cbonnett/MDN_Edward_Keras_TF)
If you could read over that would be greatly appreciated. Feel free to make adjustments, not sure I explained all the Edward stuff properly.

**Comment by dustinvtran** [comment]
This looks great—thanks for writing it up @cbonnett! I liked the visualizations. Some comments:

+ Edward views everything Bayesianly, so it could be helpful to mention that `MAP` is a method for Bayesian inference—sort of like a Bayesian version of MLE. In this case, there are no priors on parameters (that is, there is a flat prior which assigns uniform probability to all parameters). MAP, which stands for maximum a posteriori, tries to find the set of parameters which maximize the posterior distribution. For flat priors, this is equivalent to maximizing the log-likelihood. A cool thing about Edward is that by viewing this on Bayesian grounds, there are easy extensions beyond the vanilla model: for example, more complicated priors to regularize the point estimation, or perhaps something more Bayesian such as variational inference.

+ ADAM with a decaying scale factor is based off of ADVI's learning rate. It was originally added to RMSProp to enforce that the learning rate decays sufficiently fast to guarantee convergence and actually work on toy problems. We ended up using the same thing on ADAM even though ADAM technically will converge without it. (You can mention perhaps one line about this connection to ADVI, http://arxiv.org/abs/1603.00788).

> Here are all the distrbutions that are implemented in Edward:

"Here are all the distributions that are currently implemented in Edward:"
(I would add "currently"; we are sure to add more as we further develop Edward :)

> Colombia
> distrbutions

typos. "distrbutions" appears in a few places.

**Comment by cbonnett** [comment]
Thanks for the feedback, I added some extra in info on MAP and how ADAM is used and fixed some typos. I think it's ready to be published after I do a final double check. I'll post in reddit ML and statistics subreddits and via twitter, i'll be sure to mention you. 

BTW: I will be in NYC the next 3 months and 'll be attending the ICML tutorial day, any chance you will be there ?

**Comment by dustinvtran** [comment]
Yup I'll be there presenting a paper we got accepted (https://arxiv.org/abs/1511.02386). Looking forward to meeting you at ICML!

**Comment by cbonnett** [comment]
Btw, I think I might have a way of getting a proxy of the marginal likelihood for MDNs. I'll let you once I had harder think about it and it if still make sense then.

**Comment by raddy** [comment]
Any thoughts on using MDNs on higher noise problems? I've always hit a wall pretty quickly with convergence on real problems. 

An example for how quickly things can fall apart, add a scale of 25 to your r_data to quickly descend into nan exploding gradients hell.

**Comment by cbonnett** [comment]
In the example I would try and replace the Relu activation function with a tanh activation.
